src/bin/pgcopydb/ld_stream.c

```diff
@@ -767,8 +772,54 @@ streamRotateFile(LogicalStreamContext *context)
 		return true;
 	}
 
-	/* compute the WAL filename that would host the current LSN */
-	XLByteToSeg(context->cur_record_lsn, segno, context->WalSegSz);
+	/*
+	 * calculate the lsn based upon which we determine the walFileName
+	 */
+	uint64_t lsn;
+
+	/*
+	 * in most cases lsn should be same as cur_record_lsn, but occasionally few
+	 * messages of a transaction could have LSN greater than previous COMMIT
+	 * LSN. To ensure we don't switch back to previous file check if
+	 * cur_record_lsn is greater than lastWrittenLSN
+	 */
+	if (privateContext->lastWrittenLSN != InvalidXLogRecPtr)
+	{
+		/*
+		* in most cases lastWrittenLSN is less than cur_record_lsn.
+		*/
+		if (privateContext->lastWrittenLSN < context->cur_record_lsn)
+		{
+			lsn = context->cur_record_lsn;
+		}
+		/*
+		* for parallel transactions we can have messages' whose LSN is less than
+		* previous transaction's commit
+		*/
+		else if (context->cur_record_lsn < privateContext->lastWrittenLSN)
+		{
+			lsn = privateContext->lastWrittenLSN;
+		}
+		else
+		{
+			log_error("lastWrittenLSN %X/%X can't be equal to "
+					  "cur_record_lsn %X/%X",
+					  LSN_FORMAT_ARGS(privateContext->lastWrittenLSN),
+					  LSN_FORMAT_ARGS(context->cur_record_lsn));
```

Comment:
Reword message, at the moment you're saying it can't be but it is. Suggestion: 

    "Failed to compare lastWrittenLSN %X/%X with current record LSN %X/%X, expected different positions"